jest-setup.ts

```diff
@@ -0,0 +1,77 @@
+declare global {
+  namespace jest {
+    interface Matchers<R> {
+      toHaveQueryParams(expectedParamsKeys: Array<string>): R;
+      toHaveQueryParamsValues(expectedParams: Record<string, string>): R;
+      toHaveOrigin(expectedOrigin: string): R;
+    }
+  }
+}
+
+expect.extend({
+  toHaveQueryParams(received, expectedParamsKeys) {
+    const { params } = breakUrl(received);
+    for (let k of expectedParamsKeys) {
+      if (params[k] === undefined) {
+        return {
+          message: () => `URL query parameter [${k}] should exist, but it doesn't`,
+          pass: false,
+        };
+      }
+    }
+    return {
+      message: () =>
+        `URL [${received}] should not include all of the parameters ${JSON.stringify(expectedParamsKeys)}, but it does`,
+      pass: true,
+    };
+  },
+
+  toHaveQueryParamsValues(received, expectedParams) {
+    const { params } = breakUrl(received);
+    for (let k in expectedParams) {
+      if (String(params[k]) !== String(expectedParams[k])) {
+        return {
+          message: () => `URL query parameter [${k}] should have value [${expectedParams[k]}], instead it has value [${params[k]}]`,
+          pass: false,
+        };
+      }
+    }
+    return {
+      message: () =>
+        `URL [${received}] should not include all of the values [${JSON.stringify(expectedParams)}], but it does`,
+      pass: true,
+    };
+  },
+
+  toHaveOrigin(received, expectedOrigin) {
+    const { origin } = breakUrl(received);
+    if (origin !== expectedOrigin) {
+      return {
+        message: () => `URL hostname should have value [${expectedOrigin}], instead it has value [${origin}]`,
+        pass: false,
+      };
+    }
+    return {
+      message: () => `URL hostname should not have value [${expectedOrigin}], but it does`, // if .not is used
+      pass: true,
+    };
+  },
+});
+
+/* ************************ */
+
+function breakUrl(urlWithQueryParams: string) {
```

Comment:
Perhapse parseUrl would be better name for this function.

Reply:
Yes, much better, thank you.